### PR TITLE
Second part of JEP-431. Inherited Deque from SequencedCollection, fix…

### PR DESCRIPTION
**Pull request by Ihromant**
Second part of JEP-431. Inherited Deque from SequencedCollection, fixed issues in Iterator for ArrayDeque

### CODE REVIEW

File: classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java [hunk 1]
```diff
@@ -38,8 +39,8 @@ public TArrayDeque(TCollection<? extends E> c) {
         } else {
             array = new Object[c.size() + 1];
             int index = 0;
-            for (TIterator<? extends E> iter = c.iterator(); iter.hasNext();) {
-                array[index++] = iter.next();
+            for (TIterator<? extends E> it = c.iterator(); it.hasNext(); ) {
```
Comment (konsoletyper) [inline classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java:42]: What is the purpose of this change? BTW, I'm not sure if checkstyle passes extra space after last semicolon.

Comment (konsoletyper) [reply classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java:42]: BTW, this place is a good candidate to use `var`

File: classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java [hunk 2]
```diff
@@ -48,27 +49,22 @@ public TArrayDeque(TCollection<? extends E> c) {
     @Override
     public void addFirst(E e) {
         if (e == null) {
-            throw new TNullPointerException();
+            throw new NullPointerException();
```
Comment (konsoletyper) [inline classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java:52]: Perhaps, Objects.requireNonNull is better here

File: classlib/src/main/java/org/teavm/classlib/java/util/TDeque.java [hunk 3]
```diff
@@ -54,4 +43,244 @@ public interface TDeque<E> extends TQueue<E> {
     E pop();
 
     TIterator<E> descendingIterator();
+
+    @Override
+    default TDeque<E> reversed() {
+        return new ReversedDeque<>(this);
+    }
+
+    class ReversedDeque<E> implements TDeque<E> {
```
Comment (konsoletyper) [inline classlib/src/main/java/org/teavm/classlib/java/util/TDeque.java:52]: I think this class should be moved to top level with package-private visibility. Rationale: declaring class in an interface makes this class public, which can in turn have consequences with reflection.

File: classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java [hunk 4]
```diff
@@ -243,99 +233,127 @@ public boolean isEmpty() {
...
+    }
+
+    private static int modDec(int i, int mod) {
+        return --i == -1 ? mod - 1 : i;
+    }
+
     @Override
     public TIterator<E> iterator() {
-        return new TIterator<E>() {
+        return new TIterator<>() {
```
Comment (konsoletyper) [inline classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java:280]: Did you provide tests for fixes of Iterator bugs?

**Review by konsoletyper — CHANGES_REQUESTED** [review]

File: classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java (hunk 1 above)
Comment (Ihromant) [reply classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java:42]: Some "strange" changes appeared automatically in final preparation to PR. I'm just resolving IDEA issues by reformatting and selecting "obvious" fix for warning by pressing Alt-Enter.
I suppose that IDEA imported your codestyle for formatting and checkstyle will pass this change.
At least `./gradlew publishToMavenLocal` works.

File: classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java [hunk 5]
```diff
@@ -38,37 +39,28 @@ public TArrayDeque(TCollection<? extends E> c) {
         } else {
             array = new Object[c.size() + 1];
             int index = 0;
-            for (TIterator<? extends E> iter = c.iterator(); iter.hasNext();) {
-                array[index++] = iter.next();
+            for (var it = c.iterator(); it.hasNext(); ) {
```
Comment (konsoletyper) [inline classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java:42]: Chekstyle error in this and similar cases. Please, run checkstyle.

File: classlib/src/main/java/org/teavm/classlib/java/util/TReversedDeque.java [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    @Override
+    public boolean contains(Object o) {
+        return base.contains(o);
+    }
+
+    @Override
+    public Object[] toArray() {
+        Object[] arr = new Object[size()];
+        int i = 0;
+        for (TIterator<E> iter = base.descendingIterator(); iter.hasNext(); ) {
```
Comment (konsoletyper) [inline classlib/src/main/java/org/teavm/classlib/java/util/TReversedDeque.java:56]: Checkstyle error here.

**Review by konsoletyper — CHANGES_REQUESTED** [review]
